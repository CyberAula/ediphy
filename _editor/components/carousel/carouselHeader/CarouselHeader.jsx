import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from "react-redux";

import EditorIndexTitle from '../editorIndexTitle/EditorIndexTitle';
import i18n from 'i18next';
import handleCanvas from "../../../handlers/handleCanvas";
import { updateUI } from "../../../../common/actions";
import { CarouselTitleContainer, CarouselListTitle } from "./Styles";

/**
 * Carousel's header, containing the course's title and the expand/collapse buttons
 */
class CarouselHeader extends Component {
    // Initialize handlers
    hC = handleCanvas(this);

    /**
     * Render React Component
     * @returns {code}
     */
    render() {
        let courseTitle = this.props.courseTitle || i18n.t('course_title');
        let widthScroll = Math.max(courseTitle.length / 11 * 100, 100);
        // let widthScroll = this.props.courseTitle.length / 10 * 100;
        return (
            <CarouselListTitle style={{
                textAlign: this.props.carouselShow ? 'left' : 'center',
                flexFlow: this.props.carouselShow ? 'row' : 'column',
            }}
            carouselShow={this.props.carouselShow}>
                <button className="btnToggleCarousel" style={{ display: 'flex', alignItems: 'center' }} onClick={() => this.onToggleWidth()}>
                    <i style={{ fontSize: '24px', marginLeft: '4px' }} className="material-icons">{this.props.carouselShow ? 'keyboard_arrow_left' : 'menu'}</i>
                </button>
                {!this.props.carouselShow ? <br /> : null}

                <CarouselTitleContainer show={this.props.carouselShow}>
                    <EditorIndexTitle
                        id="coursetit"
                        scrollW={widthScroll}
                        title={this.props.courseTitle}
                        courseTitle
                        onNameChanged={this.hC.onTitleChanged}
                    />
                </CarouselTitleContainer>

                <div className="clear" />
            </CarouselListTitle>
        );
    }

    onToggleWidth = () => {
        if (this.props.carouselShow) {
            this.props.dispatch(updateUI({
                carouselShow: false,
                carouselFull: false,
            }));
        } else {
            this.props.dispatch(updateUI({ carouselShow: true }));
        }
    };
}

function mapStateToProps(state) {
    const { carouselFull, carouselShow } = state.reactUI;
    const courseTitle = state.undoGroup.present.globalConfig.title;
    return { carouselFull, carouselShow, courseTitle };
}

export default connect(mapStateToProps)(CarouselHeader);

CarouselHeader.propTypes = {
    /**
     * Indicates whether the index has been expanded or not
     */
    carouselShow: PropTypes.bool,
    /**
     * Title of the course as specified on the global settings
     */
    courseTitle: PropTypes.string,
    /**
     * Redux action dispatcher
     */
    // eslint-disable-next-line react/no-unused-prop-types
    dispatch: PropTypes.func.isRequired,
};
