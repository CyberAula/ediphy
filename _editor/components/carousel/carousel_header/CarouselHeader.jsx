import PropTypes from 'prop-types';
import React, { Component } from 'react';

import EditorIndexTitle from '../editor_index_title/EditorIndexTitle';

import './_carouselHeader.scss';

/**
 * Carousel's header, containing the course's title and the expand/collapse buttons
 */
export default class CarouselHeader extends Component {
    /**
     * Constructor
     * @param props
     */
    constructor(props) {
        super(props);
    }

    /**
     * Render React Component
     * @returns {code}
     */
    render() {
        return (
            <div style={{
                textAlign: this.props.carouselShow ? 'left' : 'center',
                flexFlow: this.props.carouselShow ? 'row' : 'column',
                display: 'flex',
            }}
            className={'carouselListTitle ' + (this.props.carouselShow ? 'toolbarSpread' : 'toolbarHide')}>
                <button className="btnToggleCarousel" onClick={() => this.props.onToggleWidth()}>
                    <i style={{ fontSize: this.props.carouselShow ? '16px' : '28px' }} className="material-icons">format_list_numbered</i>
                </button>
                {!this.props.carouselShow ? <br/> : null}

                <div className="navBarSpace" style={{ display: (this.props.carouselShow ? 'block' : 'none') }}>
                    <EditorIndexTitle className="tituloCurso" title={this.props.courseTitle} courseTitle onNameChanged={this.props.onTitleChanged} />
                </div>

                <button className="btnFullCarousel" style={{ position: this.props.carouselShow ? 'absolute' : 'initial' }} onClick={e => {
                    this.props.onToggleFull();
                    e.stopPropagation();
                }}>
                    <i style={{
                        fontSize: this.props.carouselShow ? '16px' : '32px',
                    }}
                    className="material-icons">{!this.props.carouselFull ? 'keyboard_arrow_right' : 'keyboard_arrow_left'}
                    </i>
                </button>

                <div className="clear" />
            </div>
        );
    }
}

CarouselHeader.propTypes = {
    /**
     * Indicates whether the index takes the whole screen's width or not
     */
    carouselFull: PropTypes.bool,
    /**
     * Indicates whether the index has been expanded or not
     */
    carouselShow: PropTypes.bool,
    /**
     * Title of the course as specified on the global settings
     */
    courseTitle: PropTypes.string,
    /**
     * Modifies the course's title
     */
    onTitleChanged: PropTypes.func.isRequired,
    /**
     * Expands the index to make it take 100% of the width
     */
    onToggleFull: PropTypes.func.isRequired,
    /**
     * Modifies the index's width
     */
    onToggleWidth: PropTypes.func.isRequired,
};
