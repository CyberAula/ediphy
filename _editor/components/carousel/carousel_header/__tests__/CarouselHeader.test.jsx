import React from 'react';
import expect from 'expect';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import CarouselHeader from '../CarouselHeader';

beforeEach(() => {
    // Forget about any mock function being called previously
    jest.clearAllMocks();
});

// Unit test's mocks
let mockFunctions = {
    onTitleChanged: jest.fn().mockName('onTitleChanged'),
    onToggleFull: jest.fn().mockName('onToggleFull'),
    onToggleWidth: jest.fn().mockName('onToggleWidth'),
};

let placeholderProps = {
    carouselFull: false,
    carouselShow: true,
    courseTitle: 'IACR',
    ...mockFunctions,
};

describe('CarouselHeader', () => {
    it('should render as expected when the carousel is expanded', () => {
        let component = shallow(<CarouselHeader {...placeholderProps}
            carouselShow />);

        expect(shallowToJson(component)).toMatchSnapshot();
    });

    it('should render as expected when the carousel is collapsed', () => {
        let component = shallow(<CarouselHeader {...placeholderProps}
            carouselShow={false} />);

        expect(shallowToJson(component)).toMatchSnapshot();
    });

    it('should show a left arrow when the carousel is at full width', () => {
        let component = shallow(<CarouselHeader {...placeholderProps}
            carouselFull />);

        expect(component.find('button.btnFullCarousel i').first().text()).toEqual('keyboard_arrow_left');
    });

    it('should handle when expand/collapse button is clicked', () => {
        let component = shallow(<CarouselHeader {...placeholderProps} />);

        component.find('button.btnToggleCarousel').simulate('click');

        expect(mockFunctions.onToggleWidth).toBeCalled();
    });

    it('should handle when the arrow is clicked', () => {
        let component = shallow(<CarouselHeader {...placeholderProps} />);

        let stopPropagation = jest.fn().mockName('stopPropagation');
        component.find('button.btnFullCarousel').simulate('click', { stopPropagation: stopPropagation });

        expect(mockFunctions.onToggleFull).toBeCalled();
        expect(stopPropagation).toBeCalled();
    });
});
