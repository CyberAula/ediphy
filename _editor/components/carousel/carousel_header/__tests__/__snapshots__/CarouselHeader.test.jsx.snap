// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarouselHeader should render as expected when the carousel is collapsed 1`] = `
<div
  className="carouselListTitle toolbarHide"
  style={
    Object {
      "display": "flex",
      "flexFlow": "column",
      "textAlign": "center",
    }
  }
>
  <button
    className="btnToggleCarousel"
    onClick={[Function]}
  >
    <i
      className="material-icons"
      style={
        Object {
          "fontSize": "28px",
        }
      }
    >
      format_list_numbered
    </i>
  </button>
  <br />
  <div
    className="navBarSpace"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <EditorIndexTitle
      className="tituloCurso"
      courseTitle={true}
      onNameChanged={[MockFunction onTitleChanged]}
      title="IACR"
    />
  </div>
  <button
    className="btnFullCarousel"
    onClick={[Function]}
    style={
      Object {
        "position": "initial",
      }
    }
  >
    <i
      className="material-icons"
      style={
        Object {
          "fontSize": "32px",
        }
      }
    >
      keyboard_arrow_right
    </i>
  </button>
  <div
    className="clear"
  />
</div>
`;

exports[`CarouselHeader should render as expected when the carousel is expanded 1`] = `
<div
  className="carouselListTitle toolbarSpread"
  style={
    Object {
      "display": "flex",
      "flexFlow": "row",
      "textAlign": "left",
    }
  }
>
  <button
    className="btnToggleCarousel"
    onClick={[Function]}
  >
    <i
      className="material-icons"
      style={
        Object {
          "fontSize": "16px",
        }
      }
    >
      format_list_numbered
    </i>
  </button>
  <div
    className="navBarSpace"
    style={
      Object {
        "display": "block",
      }
    }
  >
    <EditorIndexTitle
      className="tituloCurso"
      courseTitle={true}
      onNameChanged={[MockFunction onTitleChanged]}
      title="IACR"
    />
  </div>
  <button
    className="btnFullCarousel"
    onClick={[Function]}
    style={
      Object {
        "position": "absolute",
      }
    }
  >
    <i
      className="material-icons"
      style={
        Object {
          "fontSize": "16px",
        }
      }
    >
      keyboard_arrow_right
    </i>
  </button>
  <div
    className="clear"
  />
</div>
`;
